contracts/Rentals.sol

```diff
@@ -313,15 +274,33 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
     }
 
     /// @notice Standard function called by ERC721 contracts whenever a safe transfer occurs.
-    /// @dev The contract only allows safe transfers by itself made by the rent function.
-    /// @param _operator Caller of the safe transfer function.
+    /// Provides an alternative to acceptOffer by letting the asset holder send the asset to the contract
+    /// and accepting the offer at the same time.
+    /// IMPORTANT: Addresses that have been given allowance to an asset can safely transfer said asset to this contract
+    /// to accept an offer. The address that has been given allowance will be considered the lessor, and will enjoy all of its benefits,
+    /// including the ability to claim the asset back to themselves after the rental period is over.
+    /// @param _operator Caller of the safeTransfer function.
+    /// @param _tokenId Id of the asset received.
+    /// @param _data Bytes containing offer data.
     function onERC721Received(
         address _operator,
         address, // _from,
-        uint256, // _tokenId,
-        bytes calldata // _data
-    ) external view override returns (bytes4) {
-        require(_operator == address(this), "Rentals#onERC721Received: ONLY_ACCEPT_TRANSFERS_FROM_THIS_CONTRACT");
+        uint256 _tokenId,
+        bytes memory _data
+    ) external override returns (bytes4) {
+        if (_operator != address(this)) {
+            Offer memory offer = abi.decode(_data, (Offer));
+
+            // Check that the caller is the contract defined in the offer to ensure the function is being
+            // called through an ERC721.safeTransferFrom.
+            require(msg.sender == offer.contractAddress, "Rentals#onERC721Received: SENDER_CONTRACT_ADDRESS_MISMATCH");
```

Comment:
Maybe we can use one require here to reduce some gas

Reply:
https://github.com/decentraland/rentals-contract/pull/20